src/core/manager.c

```diff
@@ -1828,17 +1829,19 @@ static int manager_dispatch_signal_fd(sd_event_source *source, int fd, uint32_t
         }
 
         for (;;) {
-                n = read(m->signal_fd, &sfsi, sizeof(sfsi));
-                if (n != sizeof(sfsi)) {
-
-                        if (n >= 0)
-                                return -EIO;
-
-                        if (errno == EINTR || errno == EAGAIN)
+                ttl = 0;
+                do {
+                        n = read(m->signal_fd, buf + ttl, length - ttl);
+                        if (n < 0 && errno == EAGAIN) {
                                 break;
-
-                        return -errno;
-                }
+                        } else if (n < 0 && errno != EINTR) {
+                                log_error_errno(errno, "Failed to read bytes from signal file descriptor (%d): %m", errno); return -errno;
+                        } else if (n == 0) {
+                            log_error("Signal file descriptor received client disconnect, with %ld bytes discarded.", ttl); return -EIO; }
```

Comment:
The disconnect sort of must not happen; however, what if it did? There was not any logic to handle this situation. Since systemd is running with elevated privileges and is super critical to the operation of all classes of systems, it is imperative that it functions completely and correctly under all circumstances. PS: The client is the remote end of the socket - the kernel pushing wrapped signal detail messages on to the `signalfd(2)` socket file descriptor.

The problem I've experienced quite often is two-fold, with zombie processes as the end-user visible problem. When the socket has a signal buffer to read, the old code *would* partial read the message due to `errno`'s such as `EINTR` occurring. When this happens, the whole signal notification was thrown out. Additionally, any additional messages pending would then have unaligned reads, causing those (possibly full) reads to contain "junk" that further caused missed signal notifications.